File: remake_afcat_mocks.py
import json

def find_matching_bracket(text, start_idx):
    count = 0
    in_string = False
    escape = False
    
    for i in range(start_idx, len(text)):
        char = text[i]
        
        if escape:
            escape = False
            continue
            
        if char == '\\':
            escape = True
            continue
            
        if char == '"' and not escape:
            in_string = not in_string
            continue
            
        if not in_string:
            if char == '[':
                count += 1
            elif char == ']':
                count -= 1
                if count == 0:
                    return i
    return -1

with open('data.js', 'r', encoding='utf-8') as f:
    data_js = f.read()

start_marker = "CBT_EXAMS_DATABASE = "
idx = data_js.find(start_marker)
if idx == -1:
    print("Could not find CBT_EXAMS_DATABASE")
    exit(1)

array_start = data_js.find('[', idx)
array_end = find_matching_bracket(data_js, array_start)

if array_end == -1:
    print("Could not find end of array")
    exit(1)

json_str = data_js[array_start:array_end+1]

try:
    cbt_db = json.loads(json_str)
except json.JSONDecodeError as e:
    print("Failed to decode JSON:", e)
    exit(1)

original_len = len(cbt_db)
# Remove all AFCAT mocks
cbt_db = [m for m in cbt_db if m.get('exam') != 'AFCAT']
print(f"Removed {original_len - len(cbt_db)} existing AFCAT mocks.")

# Load Question Bank
try:
    with open('question_banks/cds_pyq_bank.json', 'r', encoding='utf-8') as f:
        bank = json.load(f)
except Exception as e:
    print("Failed to load question bank:", e)
    bank = {}

afcat_questions = bank.get('afcat', [])
print(f"Found {len(afcat_questions)} AFCAT questions in bank.")

for i, q in enumerate(afcat_questions):
    if not q.get('id'):
        q['id'] = f"afcat-q-{i}"
    if not q.get('topicId'):
        q['topicId'] = "mixed"

PAPER_SIZE = 100
paper_count = len(afcat_questions) // PAPER_SIZE

if paper_count == 0 and len(afcat_questions) > 0:
    paper_count = 1

for i in range(paper_count):
    start = i * PAPER_SIZE
    end = start + PAPER_SIZE
    slice_q = afcat_questions[start:end]
    if not slice_q:
        break
    
    new_mock = {
        "id": f"afcat-combined-mock-{i + 1}",
        "exam": "AFCAT",
        "subject": "Combined",
        "title": f"AFCAT Full Mock Test {i + 1}",
        "duration": 120,
        "questionsCount": len(slice_q),
        "rules": {
            "correctMarks": 3,
            "incorrectMarks": -1,
            "examType": "AFCAT"
        },
        "questions": slice_q
    }
    cbt_db.append(new_mock)

print(f"Created {paper_count} new Full Combined AFCAT mocks.")

# Rebuild data.js
new_json_str = json.dumps(cbt_db, indent=2)
new_data_js = data_js[:array_start] + new_json_str + data_js[array_end+1:]

with open('data.js', 'w', encoding='utf-8') as f:
    f.write(new_data_js)

print("data.js successfully updated!")

File: remake_afcat_mocks.js
const fs = require('fs');
const vm = require('vm');

console.log("Loading data.js...");
let dataJs = fs.readFileSync('data.js', 'utf8');

const dbEnd = dataJs.indexOf('const CURRENT_AFFAIRS_DB');
const beforeDB = dataJs.substring(0, dbEnd);
const afterDB = dataJs.substring(dbEnd);

const sandbox = {};
vm.createContext(sandbox);
vm.runInContext(beforeDB.replace('const CBT_EXAMS_DATABASE', 'var CBT_EXAMS_DATABASE'), sandbox);

let cbtDb = sandbox.CBT_EXAMS_DATABASE;

// Remove ALL existing AFCAT mocks
const originalLength = cbtDb.length;
cbtDb = cbtDb.filter(m => m.exam !== 'AFCAT');
console.log(`Removed ${originalLength - cbtDb.length} existing AFCAT mocks.`);

// Load the question bank
console.log("Loading cds_pyq_bank.json...");
const bank = JSON.parse(fs.readFileSync('question_banks/cds_pyq_bank.json', 'utf8'));

let afcatQuestions = bank.afcat || [];
console.log(`Found ${afcatQuestions.length} AFCAT questions in bank.`);

// Ensure each question has an id and topicId
afcatQuestions.forEach((q, i) => {
    if (!q.id) q.id = 'afcat-q-' + i;
    if (!q.topicId) q.topicId = 'mixed';
});

// An actual AFCAT paper has 100 questions.
const PAPER_SIZE = 100;
let paperCount = Math.floor(afcatQuestions.length / PAPER_SIZE);

if (paperCount === 0 && afcatQuestions.length > 0) {
    paperCount = 1;
}

for (let i = 0; i < paperCount; i++) {
    let start = i * PAPER_SIZE;
    let end = start + PAPER_SIZE;
    let slice = afcatQuestions.slice(start, end);
    
    // In case the last paper doesn't have 100 but we want to make it anyway
    if (slice.length === 0) break;

    cbtDb.push({
        id: `afcat-combined-mock-${i + 1}`,
        exam: "AFCAT",
        subject: "Combined",
        title: `AFCAT Full Mock Test ${i + 1}`,
        duration: 120,
        questionsCount: slice.length,
        rules: {
            correctMarks: 3,
            incorrectMarks: -1,
            examType: "AFCAT"
        },
        questions: slice
    });
}

console.log(`Created ${paperCount} new Full Combined AFCAT mocks.`);

// Reconstruct data.js
const prefixStr = dataJs.substring(0, dataJs.indexOf('const CBT_EXAMS_DATABASE = '));
const newDbStr = "const CBT_EXAMS_DATABASE = " + JSON.stringify(cbtDb, null, 2) + ";\n\n";

fs.writeFileSync('data.js', prefixStr + newDbStr + afterDB, 'utf8');
console.log("data.js successfully updated!");
